sdk/python/kfp/registry/client.py

```diff
@@ -0,0 +1,326 @@
+        url = self._config['upload_url']
+        self._refresh_creds()
+        auth = self._get_auth()
+        request_body = {}
+        if tags:
+            if isinstance(tags, str):
+                request_body = {'tags': tags}
+            elif isinstance(tags, List):
+                request_body = {'tags': ','.join(tags)}
+
+        files = {'content': open(file_name, 'rb')}
+        response = requests.post(
+            url=url,
+            data=request_body,
+            headers=extra_headers,
+            files=files,
+            auth=auth)
+        response.raise_for_status()
+        [package_name, version] = response.text.split('/')
+
+        return (package_name, version)
+
+    def _get_download_url(self,
+                          package_name: str,
+                          version: Optional[str] = None,
+                          tag: Optional[str] = None) -> str:
+        if (not version) and (not tag):
+            raise ValueError('Either version or tag must be specified.')
+        if version:
+            url = self._config['download_version_url'].format(
+                package_name=package_name, version=version)
+        if tag:
+            if version:
+                logging.info(
+                    'Both version and tag are specified, using version only.')
+            else:
+                url = self._config['download_tag_url'].format(
+                    package_name=package_name, tag=tag)
+        return url
+
+    def download_pipeline(self,
+                          package_name: str,
+                          version: Optional[str] = None,
+                          tag: Optional[str] = None,
+                          file_name: str = None) -> str:
+        url = self._get_download_url(package_name, version, tag)
+        response = self._request(request_url=url)
+
+        if not file_name:
+            file_name = package_name + '_'
+            if version:
+                file_name += version[len('sha256:'):]
+            elif tag:
+                file_name += tag
+            file_name += '.yaml'
+
+        with open(file_name, 'wb') as f:
+            f.write(response.content)
+
+        return file_name
+
+    def get_package(self, package_name: str) -> dict:
+        url = self._config['get_package_url'].format(package_name=package_name)
+        response = self._request(request_url=url)
+
+        return response.json()
+
+    def list_packages(self) -> List[dict]:
+        url = self._config['list_packages_url']
+        response = self._request(request_url=url)
+
+        return response.json()
+
+    def delete_package(self, package_name: str) -> bool:
+        url = self._config['delete_package_url'].format(
+            package_name=package_name)
+        response = self._request(request_url=url, http_request='delete')
+        response_json = response.json()
+
+        return response_json['done']
+
+    def get_version(self, package_name: str, version: str) -> dict:
```

Comment:
nit: The return type could be slightly narrowed to `Dict[str, Any]`

Reply:
Done.